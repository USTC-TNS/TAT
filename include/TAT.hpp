/* TAT.hpp
 * Copyright (C) 2019  Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef TAT_HPP_
#define TAT_HPP_

#ifndef TAT_VERSION
#define TAT_VERSION "unknown"
#endif // TAT_VERSION

#include <cassert>
#include <cstring>
#include <cmath>
#include <iostream>
#include <fstream>
#include <map>
#include <set>
#include <vector>
#include <algorithm>
#include <memory>
#include <functional>
#include <complex>

#define PASS std::cerr << "calling a passing function at " << __FILE__ << ":" << __LINE__ << " in " << __PRETTY_FUNCTION__ << std::endl, exit(233)
#define ENABLE_IF(...) class = typename std::enable_if<__VA_ARGS__::value>::type

#if (!defined TAT_USE_CPU && !defined TAT_USE_CUDA && !defined TAT_USE_DCU && !defined TAT_USE_SW)
#if !defined TAT_DEFAULT
#warning use CPU by default
#endif
#define TAT_USE_CPU
#endif

#ifdef TAT_EXTREME
#warning EXTREME compile may cost much of compile time
#endif // TAT_EXTREME

#ifdef TAT_USE_CPU
extern "C"
{
#define MKL_Complex8 std::complex<float>
#define MKL_Complex16 std::complex<double>
#include <mkl.h>
} // extern "C"
#include <hptt.h>
#ifdef TAT_EXTREME
#include <../src/hptt.cpp>
#include <../src/plan.cpp>
#include <../src/transpose.cpp>
#include <../src/utils.cpp>
#endif // TAT_EXTREME
#include <rang.hpp>

// SVD
#if (defined TAT_USE_GESDD && defined TAT_USE_GESVD) || (defined TAT_USE_GESVD && defined TAT_USE_GESVDX) || (defined TAT_USE_GESVDX && defined TAT_USE_GESDD)
#error only one of GESDD, GESVD and GESVDX could be in use
#endif
#if (!defined TAT_USE_GESDD && !defined TAT_USE_GESVD && !defined TAT_USE_GESVDX)
#if !defined TAT_DEFAULT
#warning must use one of GESDD, GESVD and GESVDX, default use GESVD now
#endif
#define TAT_USE_GESVD
#endif

// QR
#if (defined TAT_USE_GEQRF && defined TAT_USE_GEQP3)
#error only one of GEQRF and GEQP3 could be in use
#endif
#if (!defined TAT_USE_GEQRF && !defined TAT_USE_GEQP3)
#if !defined TAT_DEFAULT
#warning must use one of GEQRF and GEQP3, default use GEQRF now
#endif
#define TAT_USE_GEQRF
#endif

#ifdef TAT_USE_GEQP3
#error GEQP3 is current unusable
#endif

#endif // TAT_USE_CPU

namespace TAT {
  template<class T>
  class is_scalar {
   public:
    static constexpr bool value = std::is_scalar<T>::value || std::is_same<T, std::complex<float>>::value || std::is_same<T, std::complex<double>>::value || std::is_same<T, std::complex<long double>>::value;
  };

  template<class Base>
  class RealBaseClass {
   public:
    using type=Base;
  };
  template<>
  class RealBaseClass<std::complex<float>> {
   public:
    using type=float;
  };
  template<>
  class RealBaseClass<std::complex<double>> {
   public:
    using type=double;
  };
  template<>
  class RealBaseClass<std::complex<long double>> {
   public:
    using type=long double;
  };
  template<class T>
  using RealBase = typename RealBaseClass<T>::type;

  enum class Device : unsigned char {CPU, CUDA, DCU, SW};

  namespace legs {
    class Legs {
     public:
      unsigned char id;
      bool operator==(const Legs& other) const {
        return id==other.id;
      }
      bool operator!=(const Legs& other) const {
        return id!=other.id;
      }
      bool operator<(const Legs& other) const {
        return id<other.id;
      }
      Legs() = default;
      Legs(const std::string& name) {
        try {
          id = name2id.at(name);
        } catch (const std::out_of_range& e) {
          id = total++;
          name2id[name]=id;
          id2name[id]=name;
        } // exsit name
      }
      static unsigned char total;
      static std::map<std::string, unsigned char> name2id;
      static std::map<unsigned char, std::string> id2name;

      friend std::ostream& operator<<(std::ostream& out, const Legs& value) {
        try {
          return out << id2name.at(value.id);
        } catch (const std::out_of_range& e) {
          return out << "UserDefinedLeg" << value.id;
        } // get name
      } // operator<<
    }; // class Legs

    unsigned char Legs::total = 0;
    std::map<std::string, unsigned char> Legs::name2id = {};
    std::map<unsigned char, std::string> Legs::id2name = {};
  } // namespace legs
  using legs::Legs;

  namespace legs_name {
#define TAT_DefineLeg(x) static const TAT::Legs x(#x)
#define TAT_DefineLegs(n) \
      TAT_DefineLeg(Phy##n); TAT_DefineLeg(Left##n); TAT_DefineLeg(Right##n); TAT_DefineLeg(Up##n); TAT_DefineLeg(Down##n); \
      TAT_DefineLeg(LeftUp##n); TAT_DefineLeg(LeftDown##n); TAT_DefineLeg(RightUp##n); TAT_DefineLeg(RightDown##n)
#define TAT_Legs \
      TAT_DefineLegs(); TAT_DefineLegs(1); TAT_DefineLegs(2); TAT_DefineLegs(3); TAT_DefineLegs(4); \
      TAT_DefineLegs(5); TAT_DefineLegs(6); TAT_DefineLegs(7); TAT_DefineLegs(8); TAT_DefineLegs(9)
    TAT_Legs;
#undef TAT_Legs
#undef TAT_DefineLegs
#define TAT_DefineLegs(n) \
      TAT_DefineLeg(Tmp##n##0); TAT_DefineLeg(Tmp##n##1); TAT_DefineLeg(Tmp##n##2); TAT_DefineLeg(Tmp##n##3); TAT_DefineLeg(Tmp##n##4); \
      TAT_DefineLeg(Tmp##n##5); TAT_DefineLeg(Tmp##n##6); TAT_DefineLeg(Tmp##n##7); TAT_DefineLeg(Tmp##n##8); TAT_DefineLeg(Tmp##n##9)
#define TAT_Legs \
      TAT_DefineLegs(); TAT_DefineLegs(1); TAT_DefineLegs(2); TAT_DefineLegs(3); TAT_DefineLegs(4); \
      TAT_DefineLegs(5); TAT_DefineLegs(6); TAT_DefineLegs(7); TAT_DefineLegs(8); TAT_DefineLegs(9)
    TAT_Legs;
#undef TAT_Legs
#undef TAT_DefineLegs
#undef TAT_DefineLeg
    // total = 190 currently
  } // namespace legs_name

  using Size = std::size_t;
  using Rank = unsigned int;

  namespace data {
    template<Device device, class Base>
    class Magic;
#define DefineData(x) \
      namespace x { \
        template<class Base> \
        class Data; \
      } \
      template<class Base> \
      class Magic<Device::x, Base>{ \
       public: \
        using type=x::Data<Base>; \
      }

    DefineData(CPU);
    DefineData(CUDA);
    DefineData(DCU);
    DefineData(SW);
#undef DefineData

    template<Device device=Device::CPU, class Base=double, ENABLE_IF(is_scalar<Base>)>
    using Data = typename Magic<device, Base>::type;
  } // namespace data
  using data::Data;

  namespace block {
    template<Device device=Device::CPU, class Base=double>
    class Block;
  } // namespace block
  using block::Block;

  namespace tensor {
    template<Device device=Device::CPU, class Base=double>
    class Tensor;
  } // namespace tensor
  //using tensor::Tensor;
  template<Device device=Device::CPU, class Base=double>
  using Tensor=Block<device, Base>;

  namespace node {
    template<Device device=Device::CPU, class Base=double>
    class Node;
  } // namespace node
  using node::Node;

  namespace lazy {
    template<Device device=Device::CPU, class Base=double>
    class Lazy;
  } // namespace lazy
  using lazy::Lazy;

  namespace site {
    template<Device device=Device::CPU, class Base=double>
    class Site;
  } // namespace site
  using site::Site;

  namespace lattice {
    template<int dimension>
    class Dimension;

    template<class Tags=Dimension<2>, Device device=Device::CPU, class Base=double>
    class Lattice;
  } // namespace lattice
  using site::Site;
} // namespace TAT

#include "TAT/Data.hpp"
#include "TAT/Block.hpp"
// Tensor
#include "TAT/Node.hpp"
#include "TAT/Lazy.hpp"

#include "TAT/Site.hpp"
#include "TAT/Lattice.hpp"

#endif // TAT_HPP_
