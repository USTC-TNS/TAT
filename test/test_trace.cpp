/**
 * Copyright (C) 2019-2021 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <TAT/TAT.hpp>

#include "run_test.hpp"

#define t_edge(...) \
   { {__VA_ARGS__}, true }
#define f_edge(...) \
   { {__VA_ARGS__}, false }

void run_test() {
   std::cout << TAT::Tensor<double, TAT::NoSymmetry>({"A", "B", "C", "D", "E"}, {2, 3, 2, 3, 4}).range().trace({{"A", "C"}, {"B", "D"}}) << "\n";
   std::cout << TAT::Tensor<double, TAT::NoSymmetry>({"A", "B", "C"}, {2, 2, 3}).range().trace({{"A", "B"}}) << "\n";
   auto a = TAT::Tensor<double, TAT::NoSymmetry>({"A", "B", "C"}, {4, 3, 5}).range();
   auto b = TAT::Tensor<double, TAT::NoSymmetry>({"D", "E", "F"}, {5, 4, 6}).range();
   std::cout << a.contract(b, {{"A", "E"}, {"C", "D"}}) - a.contract(b, {}).trace({{"A", "E"}, {"C", "D"}}) << "\n";

   do {
      std::cout << "\n\n#\n";
      auto c =
            TAT::Tensor<double, TAT::FermiSymmetry>{
                  {"A", "B", "C"},
                  {t_edge({-1, 1}, {0, 1}, {-2, 1}), f_edge({0, 1}, {1, 2}), f_edge({0, 2}, {1, 2})}}
                  .range();
      auto d =
            TAT::Tensor<double, TAT::FermiSymmetry>{
                  {"E", "F", "G"},
                  {f_edge({0, 2}, {1, 1}), t_edge({-2, 1}, {-1, 1}, {0, 2}), t_edge({0, 1}, {-1, 2})}}
                  .range();
      std::cout << c << "\n";
      std::cout << d << "\n";
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c, d, {{"B", "G"}}) << "\n";
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c.transpose({"A", "C", "B"}), d.transpose({"G", "E", "F"}), {{"B", "G"}})
                << "\n";
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c, d, {}).trace({{"B", "G"}}) << "\n";
      std::cout
            << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c.transpose({"A", "C", "B"}), d.transpose({"G", "E", "F"}), {}).trace({{"B", "G"}})
            << "\n";
   } while (false);
   do {
      std::cout << "\n\n#\n";
      auto c =
            TAT::Tensor<double, TAT::FermiSymmetry>{
                  {"A", "C", "D"},
                  {t_edge({-1, 1}, {0, 1}, {-2, 1}), f_edge({0, 2}, {1, 2}), t_edge({0, 2}, {-1, 1}, {-2, 2})}}
                  .range();
      auto d =
            TAT::Tensor<double, TAT::FermiSymmetry>{
                  {"E", "F", "H"},
                  {f_edge({0, 2}, {1, 1}), t_edge({-2, 1}, {-1, 1}, {0, 2}), f_edge({0, 2}, {1, 1}, {2, 2})}}
                  .range();
      std::cout << c << "\n";
      std::cout << d << "\n";
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c, d, {{"D", "H"}}) << "\n";
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c.transpose({"A", "C", "D"}), d.transpose({"H", "E", "F"}), {{"D", "H"}})
                << "\n";
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c, d, {}).trace({{"D", "H"}}) << "\n";
      std::cout
            << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c.transpose({"A", "C", "D"}), d.transpose({"H", "E", "F"}), {}).trace({{"D", "H"}})
            << "\n";
   } while (false);
   do {
      std::cout << "\n\n#\n";
      auto c =
            TAT::Tensor<double, TAT::U1Symmetry>{
                  {"A", "B", "C", "D"},
                  {t_edge({-1, 1}, {0, 1}, {-2, 1}), f_edge({0, 1}, {1, 2}), f_edge({0, 2}, {1, 2}), t_edge({0, 2}, {-1, 1}, {-2, 2})}}
                  .range();
      auto d =
            TAT::Tensor<double, TAT::U1Symmetry>{
                  {"E", "F", "G", "H"},
                  {f_edge({0, 2}, {1, 1}), t_edge({-2, 1}, {-1, 1}, {0, 2}), t_edge({0, 1}, {-1, 2}), f_edge({0, 2}, {1, 1}, {2, 2})}}
                  .range();
      std::cout << TAT::Tensor<double, TAT::U1Symmetry>::contract(c, d, {{"B", "G"}, {"D", "H"}}) << "\n";
      std::cout << TAT::Tensor<double, TAT::U1Symmetry>::contract(
                         c.transpose({"A", "C", "B", "D"}),
                         d.transpose({"G", "H", "E", "F"}),
                         {{"B", "G"}, {"D", "H"}})
                << "\n";
      std::cout << TAT::Tensor<double, TAT::U1Symmetry>::contract(c, d, {}).trace({{"B", "G"}}).trace({{"D", "H"}}) << "\n";
      std::cout << TAT::Tensor<double, TAT::U1Symmetry>::contract(c, d, {}).trace({{"B", "G"}, {"D", "H"}}) << "\n";
      std::cout << TAT::Tensor<double, TAT::U1Symmetry>::contract(c.transpose({"A", "C", "B", "D"}), d.transpose({"G", "H", "E", "F"}), {})
                         .trace({{"B", "G"}, {"D", "H"}})
                << "\n";
   } while (false);
   do {
      std::cout << "\n\n#\n";
      auto c =
            TAT::Tensor<double, TAT::FermiSymmetry>{
                  {"A", "B", "C", "D"},
                  {t_edge({-1, 1}, {0, 1}, {-2, 1}), f_edge({0, 1}, {1, 2}), f_edge({0, 2}, {1, 2}), t_edge({-2, 2}, {-1, 1}, {0, 2})}}
                  .range();
      auto d =
            TAT::Tensor<double, TAT::FermiSymmetry>{
                  {"E", "F", "G", "H"},
                  {f_edge({0, 2}, {1, 1}), t_edge({-2, 1}, {-1, 1}, {0, 2}), t_edge({0, 1}, {-1, 2}), f_edge({2, 2}, {1, 1}, {0, 2})}}
                  .range();
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c, d, {{"B", "G"}, {"D", "H"}}) << "\n";
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(
                         c.transpose({"A", "C", "B", "D"}),
                         d.transpose({"G", "H", "E", "F"}),
                         {{"B", "G"}, {"D", "H"}})
                << "\n";
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c, d, {}).trace({{"B", "G"}}).trace({{"D", "H"}}) << "\n";
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c, d, {}).trace({{"B", "G"}, {"D", "H"}}) << "\n";
      std::cout << TAT::Tensor<double, TAT::FermiSymmetry>::contract(c.transpose({"A", "C", "B", "D"}), d.transpose({"G", "H", "E", "F"}), {})
                         .trace({{"B", "G"}, {"D", "H"}})
                << "\n";
   } while (false);
}
