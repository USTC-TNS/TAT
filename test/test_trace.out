{[32mnames[0m:[E],[32medges[0m:[4],[32mblocks[0m:[420,426,432,438]}
{[32mnames[0m:[C],[32medges[0m:[3],[32mblocks[0m:[9,11,13]}
{[32mnames[0m:[B,F],[32medges[0m:[3,6],[32mblocks[0m:[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}


#
{[32mnames[0m:[A,B,C],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:1,1:2}},{arrow:0,segment:{0:2,1:2}}],[32mblocks[0m:{[33m[-1,0,1][0m:[0,1],[33m[-1,1,0][0m:[2,3,4,5],[33m[0,0,0][0m:[6,7],[33m[-2,1,1][0m:[8,9,10,11]}}
{[32mnames[0m:[E,F,G],[32medges[0m:[{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}},{arrow:1,segment:{0:1,-1:2}}],[32mblocks[0m:{[33m[0,0,0][0m:[0,1,2,3],[33m[1,-1,0][0m:[4],[33m[1,0,-1][0m:[5,6,7,8]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[34,46,45,61],[33m[-1,1,0,0][0m:[0,0,0,0,0,1,2,3],[33m[-1,1,1,-1][0m:[0,4],[33m[0,0,0,0][0m:[0,6,12,18,0,7,14,21],[33m[0,0,1,-1][0m:[24,28],[33m[0,1,0,-1][0m:[0,0,0,0],[33m[0,1,1,-2][0m:[0,0],[33m[-2,1,1,0][0m:[-100,-136,-111,-151]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[34,46,45,61],[33m[-1,1,0,0][0m:[0,0,0,0,0,1,2,3],[33m[-1,1,1,-1][0m:[0,4],[33m[0,0,0,0][0m:[0,6,12,18,0,7,14,21],[33m[0,0,1,-1][0m:[24,28],[33m[0,1,0,-1][0m:[0,0,0,0],[33m[0,1,1,-2][0m:[0,0],[33m[-2,1,1,0][0m:[-100,-136,-111,-151]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[34,46,45,61],[33m[-1,1,0,0][0m:[0,0,0,0,0,1,2,3],[33m[-1,1,1,-1][0m:[0,4],[33m[0,0,0,0][0m:[0,6,12,18,0,7,14,21],[33m[0,0,1,-1][0m:[24,28],[33m[0,1,0,-1][0m:[0,0,0,0],[33m[0,1,1,-2][0m:[0,0],[33m[-2,1,1,0][0m:[-100,-136,-111,-151]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[34,46,45,61],[33m[-1,1,0,0][0m:[0,0,0,0,0,1,2,3],[33m[-1,1,1,-1][0m:[0,4],[33m[0,0,0,0][0m:[0,6,12,18,0,7,14,21],[33m[0,0,1,-1][0m:[24,28],[33m[0,1,0,-1][0m:[0,0,0,0],[33m[0,1,1,-2][0m:[0,0],[33m[-2,1,1,0][0m:[-100,-136,-111,-151]}}


#
{[32mnames[0m:[A,C,D],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:1,segment:{0:2,-1:1,-2:2}}],[32mblocks[0m:{[33m[-1,1,0][0m:[0,1,2,3],[33m[0,0,0][0m:[4,5,6,7],[33m[0,1,-1][0m:[8,9]}}
{[32mnames[0m:[E,F,H],[32medges[0m:[{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}},{arrow:0,segment:{0:2,1:1,2:2}}],[32mblocks[0m:{[33m[0,-2,2][0m:[0,1,2,3],[33m[0,-1,1][0m:[4,5],[33m[0,0,0][0m:[6,7,8,9,10,11,12,13],[33m[1,-2,1][0m:[14],[33m[1,-1,0][0m:[15,16]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,0,0,0],[33m[-1,1,0,0][0m:[7,9,11,13,33,43,53,63],[33m[-1,1,1,-1][0m:[16,78],[33m[0,0,0,0][0m:[59,77,95,113,85,111,137,163],[33m[0,0,1,-1][0m:[140,202],[33m[0,1,0,-1][0m:[-32,-40,-36,-45],[33m[0,1,1,-2][0m:[-112,-126],[33m[-2,1,1,0][0m:[0,0,0,0]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,0,0,0],[33m[-1,1,0,0][0m:[7,9,11,13,33,43,53,63],[33m[-1,1,1,-1][0m:[16,78],[33m[0,0,0,0][0m:[59,77,95,113,85,111,137,163],[33m[0,0,1,-1][0m:[140,202],[33m[0,1,0,-1][0m:[-32,-40,-36,-45],[33m[0,1,1,-2][0m:[-112,-126],[33m[-2,1,1,0][0m:[0,0,0,0]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,0,0,0],[33m[-1,1,0,0][0m:[7,9,11,13,33,43,53,63],[33m[-1,1,1,-1][0m:[16,78],[33m[0,0,0,0][0m:[59,77,95,113,85,111,137,163],[33m[0,0,1,-1][0m:[140,202],[33m[0,1,0,-1][0m:[-32,-40,-36,-45],[33m[0,1,1,-2][0m:[-112,-126],[33m[-2,1,1,0][0m:[0,0,0,0]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[0,0,0,0],[33m[-1,1,0,0][0m:[7,9,11,13,33,43,53,63],[33m[-1,1,1,-1][0m:[16,78],[33m[0,0,0,0][0m:[59,77,95,113,85,111,137,163],[33m[0,0,1,-1][0m:[140,202],[33m[0,1,0,-1][0m:[-32,-40,-36,-45],[33m[0,1,1,-2][0m:[-112,-126],[33m[-2,1,1,0][0m:[0,0,0,0]}}


#
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{-1:1,0:1,-2:1},{0:2,1:2},{0:2,1:1},{-2:1,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[601,655,709,763,704,770,836,902],[33m[-1,1,1,-1][0m:[1012,1229],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[944,1420,1008,1517],[33m[0,1,1,-2][0m:[4314,4604],[33m[-2,1,1,0][0m:[5922,6506,6246,6862]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{-1:1,0:1,-2:1},{0:2,1:2},{0:2,1:1},{-2:1,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[601,655,709,763,704,770,836,902],[33m[-1,1,1,-1][0m:[1012,1229],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[944,1420,1008,1517],[33m[0,1,1,-2][0m:[4314,4604],[33m[-2,1,1,0][0m:[5922,6506,6246,6862]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{-1:1,0:1,-2:1},{0:2,1:2},{0:2,1:1},{-2:1,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[601,655,709,763,704,770,836,902],[33m[-1,1,1,-1][0m:[1012,1229],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[944,1420,1008,1517],[33m[0,1,1,-2][0m:[4314,4604],[33m[-2,1,1,0][0m:[5922,6506,6246,6862]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{-1:1,0:1,-2:1},{0:2,1:2},{0:2,1:1},{-2:1,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[601,655,709,763,704,770,836,902],[33m[-1,1,1,-1][0m:[1012,1229],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[944,1420,1008,1517],[33m[0,1,1,-2][0m:[4314,4604],[33m[-2,1,1,0][0m:[5922,6506,6246,6862]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{-1:1,0:1,-2:1},{0:2,1:2},{0:2,1:1},{-2:1,-1:1,0:2}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[601,655,709,763,704,770,836,902],[33m[-1,1,1,-1][0m:[1012,1229],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[944,1420,1008,1517],[33m[0,1,1,-2][0m:[4314,4604],[33m[-2,1,1,0][0m:[5922,6506,6246,6862]}}


#
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[-571,-621,-671,-721,-558,-604,-650,-696],[33m[-1,1,1,-1][0m:[-940,-873],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[-944,-1420,-1008,-1517],[33m[0,1,1,-2][0m:[-4314,-4604],[33m[-2,1,1,0][0m:[-5922,-6506,-6246,-6862]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[-571,-621,-671,-721,-558,-604,-650,-696],[33m[-1,1,1,-1][0m:[-940,-873],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[-944,-1420,-1008,-1517],[33m[0,1,1,-2][0m:[-4314,-4604],[33m[-2,1,1,0][0m:[-5922,-6506,-6246,-6862]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[-571,-621,-671,-721,-558,-604,-650,-696],[33m[-1,1,1,-1][0m:[-940,-873],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[-944,-1420,-1008,-1517],[33m[0,1,1,-2][0m:[-4314,-4604],[33m[-2,1,1,0][0m:[-5922,-6506,-6246,-6862]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[-571,-621,-671,-721,-558,-604,-650,-696],[33m[-1,1,1,-1][0m:[-940,-873],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[-944,-1420,-1008,-1517],[33m[0,1,1,-2][0m:[-4314,-4604],[33m[-2,1,1,0][0m:[-5922,-6506,-6246,-6862]}}
{[32mnames[0m:[A,C,E,F],[32medges[0m:[{arrow:1,segment:{-1:1,0:1,-2:1}},{arrow:0,segment:{0:2,1:2}},{arrow:0,segment:{0:2,1:1}},{arrow:1,segment:{-2:1,-1:1,0:2}}],[32mblocks[0m:{[33m[-1,0,1,0][0m:[1062,1166,1386,1522],[33m[-1,1,0,0][0m:[-571,-621,-671,-721,-558,-604,-650,-696],[33m[-1,1,1,-1][0m:[-940,-873],[33m[0,0,0,0][0m:[1515,1673,1831,1989,1618,1788,1958,2128],[33m[0,0,1,-1][0m:[2898,3115],[33m[0,1,0,-1][0m:[-944,-1420,-1008,-1517],[33m[0,1,1,-2][0m:[-4314,-4604],[33m[-2,1,1,0][0m:[-5922,-6506,-6246,-6862]}}
